src/Tizen.Network.Stc/Tizen.Network.Stc/StcManagerImpl.cs

```diff
@@ -0,0 +1,206 @@
+        private int _requestId = 0;
+        private int _requestAllId = 0;
+
+        private StcManagerImpl()
+        {
+        }
+
+        internal static StcManagerImpl Instance
+        {
+            get
+            {
+                if (_instance == null)
+                {
+                    Log.Debug(Globals.LogTag, "Instance is null");
+                    _instance = new StcManagerImpl();
+                }
+
+                return _instance;
+            }
+        }
+
+        private static ThreadLocal<StcHandleHolder> s_threadName = new ThreadLocal<StcHandleHolder>(() =>
+        {
+            Log.Info(Globals.LogTag, "In threadlocal delegate");
+            return new StcHandleHolder();
+        });
+
+        internal SafeStcHandle GetSafeHandle()
+        {
+            return s_threadName.Value.GetSafeHandle();
+        }
+
+        internal SafeStcHandle Initialize()
+        {
+            SafeStcHandle handle;
+            int ret = Interop.Stc.Initialize(out handle);
+            if (ret != (int)StcError.None)
+            {
+                Log.Error(Globals.LogTag, "Failed to initialize Stc, Error - " + (StcError)ret);
+                StcErrorFactory.ThrowStcException(ret);
+            }
+            return handle;
+        }
+
+        internal Task<NetworkStatistics> GetStatisticsAsync(string appId, StcRule rule)
+        {
+            TaskCompletionSource<NetworkStatistics> task = new TaskCompletionSource<NetworkStatistics>();
+            IntPtr id;
+
+            lock (_getStatsCb_map)
+            {
+                id = (IntPtr)_requestId++;
+                _getStatsCb_map[id] = (int result, Interop.Stc.SafeStatsHandle info, IntPtr key) =>
+                {
+                    if(result != (int)StcError.None)
+                    {
+                        Log.Error(Globals.LogTag, "GetStats failed, Error - " + (StcError)result);
+                        task.SetException(new InvalidOperationException("Error occurs during GetStats(), " + (StcError)result));
+                    }
+
+                    Interop.Stc.SafeStatsHandle cloned;
+                    int retValue = Interop.Stc.Info.StatsClone(info, out cloned);
+                    if (retValue != (int)StcError.None)
+                    {
+                        Log.Error(Globals.LogTag, "StatsClone() failed , Error - " + (StcError)retValue);
+                        task.SetException(new InvalidOperationException("Error occurs during StatsClone(), " + (StcError)retValue));
+                    }
+
+                    task.SetResult(new NetworkStatistics(cloned));
+                    lock (_getStatsCb_map)
+                    {
+                        _getStatsCb_map.Remove(key);
+                    }
+
+                    return CallbackRet.Continue;
+                };
+            }
+
+            int ret = Interop.Stc.GetStats(GetSafeHandle(), rule.ConvertToNativeRule(appId), _getStatsCb_map[id], id);
```

Comment:
I recommend to release SafeRuleHandle after use

``` c#
using (var ruleHandle = rule.ConvertToNativeRule(appId)) 
{
     int ret = Interop.Stc.GetStats(GetSafeHandle(), ruleHandle , _getStatsCb_map[id], id);
    ....
}

```

Reply:
Done as suggested.
Earlier I thought SafeHandle will take care of it, and dispose the ruleHandle when reference count is zero.